include/zephyr/linker/common-rom/common-rom-kernel-devices.ld

```diff
@@ -22,6 +22,10 @@
 
 	ITERABLE_SECTION_ROM_NUMERIC(device, 4)
 
+#ifdef CONFIG_DEVICE_DEFER_INIT
+	ITERABLE_SECTION_ROM(z_device_before_init, 4)
```

Comment:
cool! you should also add an entry in `cmake/linker_script/common/common-rom.cmake` for platforms that use `CONFIG_CMAKE_LINKER_GENERATOR`